utils/frame/remote-externalities/src/lib.rs

```diff
@@ -62,7 +62,8 @@ const DEFAULT_HTTP_ENDPOINT: &str = "https://rpc.polkadot.io:443";
 struct Snapshot<B: BlockT> {
```

Comment:
To keep compatibility we can just add:
`snapshot_version: u32` that start at 2 (State version can currently be 0 or 1) , and then return the error as you suggest (having a version is usally a good idea to me, even if here it would just be for the error message).
maybe more `Compact<u32>` (not too sure how the state version is currently encoded).

Reply:
Wouldn't Decode::decode fail if the internal structure of the Snapshot is changed, so we wouldn't be able to check the snapshot version?

I was thinking about implementing something like this, where we decode it in two layers

```rust
// First layer we get the snapshot version and the inner bytes of the snapshot
#[derive(Decode, Encode)]
struct Snapshot<B: BlockT> {
	version: SnapshotVersion,
        inner: Bytes
}

// Actual snapshot we decode after checking the version
#[derive(Decode, Encode)]
struct SnapshotInner<B: BlockT> {
	state_version: StateVersion,
	block_hash: B::Hash,
	// <Vec<Key, (Value, MemoryDbRefCount)>>
	raw_storage: Vec<(H256, (Vec<u8>, i32))>,
	storage_root: H256,
}
```

but felt it might be overkill. Let me know what you think

Reply:
struct is just a sequence of it s field encoded. So the enum for stateversion probably currently encode to either 0 or 1 (I fear it may be one a single byte), if we encode the snapshot version in a similar way and start versioning at 2, it should be fine.

Reply:
Ah, I see now that I should be able to decode only the first field using something like

```rust
pub struct PartialSnapshot {
    pub first_field: u32,
}
```

I'll try this method for adding a snapshot version (starting versioning at > than what the StateVersion could be as you suggest)

Reply:
I mean 
```
struct A {
  a: string,
  struct B {
    b: u8,
  }
}
```
encode the same way as:
```
struct C {
  c: string,
  d: u8,
}
```

Reply:
> I pushed a commit adding snapshot versioning. Is that about what you were imagining?

yes, just need to ensure StateVersion and u16 > 1 cannot encode to the same thing. (alternatively can use Compact<u16>).

Reply:
StateVersion is an enum with 2 variants (I think meaning they are encoded either 0 or 1). 

So I think should be fine for any snapshots created prior to this change?

Reply:
Just checked, V0 will encode to [0] ,v1 to [1]. so with u8 as version it works. with u16 it does not (u16 encode to 2bytes le). with 16 it all depend on the value of byte number 2 so it is a bit awkward.
But just Compact<u16> or any compact integer would be fine. The only issue is that [0] will decode to 0 but [1] will error (out of range error, this is an invalid encoding). But tbh, as long as we got an error I think it is fine.